gluten-data/src/main/scala/io/glutenproject/vectorized/GlutenColumnarBatchSerializer.scala

```diff
@@ -67,8 +65,16 @@ private class GlutenColumnarBatchSerializerInstance(schema: StructType,
           SparkSchemaUtil.toArrowSchema(schema, SQLConf.get.sessionLocalTimeZone)
         GlutenArrowAbiUtil.exportSchema(allocator, arrowSchema, cSchema)
         val handle = ShuffleReaderJniWrapper.make(
-          JniByteInputStreams.create(in), cSchema.memoryAddress())
+          JniByteInputStreams.create(in), cSchema.memoryAddress(),
+          NativeMemoryAllocators.contextInstance.getNativeInstanceId)
         cSchema.close()
+        // Close shuffle reader instance as lately as the end of task processing,
+        // since the native reader could hold a reference to memory pool that
+        // was used to create all buffers read from shuffle reader. The pool
+        // should keep alive before all buffers finish consuming.
+        TaskMemoryResources.addRecycler(50) { _ =>
+          ShuffleReaderJniWrapper.close(handle)
```

Comment:
If we just close the `ShuffleReaderJniWrapper` at the end of the task, we do not release this instance as soon as possible.

The `arrow::schema` may exist as a very long lifecycle.

Maybe we could release other members except pool as soon as possible?

Reply:
> The arrow::schema may exist as a very long lifecycle.

I think this doesn't really matter since the schema just holds a little of memory space.

> Maybe we could release other members except pool as soon as possible?

I was thinking about it but that requires for some arch-level changes. Let's just solve the issues with a minimal change then start a new topic about that.

Reply:
I'll add some comments to the API in https://github.com/oap-project/gluten/pull/1527